### PR TITLE
[SPARK-44340][SQL] Define the computing logic through PartitionEvaluator API and use it in WindowGroupLimitExec

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
`WindowGroupLimitExec` is updated to use the PartitionEvaluator API to do execution.

### Why are the changes needed?
To define the computing logic and requires the caller side to explicitly list what needs to be serialized and sent to executors

### Does this PR introduce _any_ user-facing change?
'No'.
Just update the inner implementation.

### How was this patch tested?
Test cases updated & running benchmark manually.

```
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Benchmark Top-K:                                                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] -----------------------------------------------------------------------------------------------------------------------------------------------
[info] ROW_NUMBER (PARTITION: , WindowGroupLimit: false)                        10622          11266         617          2.0         506.5       1.0X
[info] ROW_NUMBER (PARTITION: , WindowGroupLimit: true)                          1712           1744          19         12.2          81.6       6.2X
[info] ROW_NUMBER (PARTITION: PARTITION BY b, WindowGroupLimit: false)          23679          25107         NaN          0.9        1129.1       0.4X
[info] ROW_NUMBER (PARTITION: PARTITION BY b, WindowGroupLimit: true)            6381           6527          95          3.3         304.3       1.7X
[info] RANK (PARTITION: , WindowGroupLimit: false)                              11492          11631         106          1.8         548.0       0.9X
[info] RANK (PARTITION: , WindowGroupLimit: true)                                2675           2920         118          7.8         127.5       4.0X
[info] RANK (PARTITION: PARTITION BY b, WindowGroupLimit: false)                24208          24299          95          0.9        1154.3       0.4X
[info] RANK (PARTITION: PARTITION BY b, WindowGroupLimit: true)                  6347           6478          85          3.3         302.6       1.7X
[info] DENSE_RANK (PARTITION: , WindowGroupLimit: false)                        11288          11959         458          1.9         538.2       0.9X
[info] DENSE_RANK (PARTITION: , WindowGroupLimit: true)                          2684           2945         144          7.8         128.0       4.0X
[info] DENSE_RANK (PARTITION: PARTITION BY b, WindowGroupLimit: false)          24316          25130         711          0.9        1159.5       0.4X
[info] DENSE_RANK (PARTITION: PARTITION BY b, WindowGroupLimit: true)            6589           6925         383          3.2         314.2       1.6X
```

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan cc @vinodkc

**Comment by beliefer** [comment]
cc @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitEvaluatorFactory.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+          input => DenseRankLimitIterator(childOutput, input, orderSpec, limit, numOutputRows))
+      case _: DenseRank =>
+        new WindowGroupLimitPartitionEvaluator(
+          input => new GroupedLimitIterator(input, childOutput, partitionSpec,
+            (input: Iterator[InternalRow]) =>
+              DenseRankLimitIterator(childOutput, input, orderSpec, limit, numOutputRows)))
+    }
+
+  }
+
```
Comment (vinodkc) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitEvaluatorFactory.scala:62]: Can we create the function in the match and then call `new WindowGroupLimitPartitionEvaluator(f) outside the match expression?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitEvaluatorFactory.scala:53]: Done.

**Comment by viirya** [comment]
The CI failure looks unrelated.

**Review by viirya — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan @viirya @vinodkc Thank you!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala [hunk 2]
```diff
@@ -73,26 +73,23 @@ case class WindowGroupLimitExec(
...
+        limit,
+        child.output,
+        numOutputRows)
+
+    if (conf.usePartitionEvaluator) {
+      child.execute().mapPartitionsWithEvaluator(evaluatorFactory)
+    } else {
+      child.execute().mapPartitionsInternal { iter =>
+        val evaluator = evaluatorFactory.createEvaluator()
+        evaluator.eval(0, iter)
```
Comment (vinodkc) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala:91]: @beliefer, Can you please raise a follow-up PR to handle the partition index as this https://github.com/apache/spark/pull/42185

**Pull request closed without merging**